### PR TITLE
[enhancement] add serialization option that inserts final newline

### PR DESCRIPTION
**Pull request by line-o**
Introduce new serialization option `exist:insert-final-newline` for XML _and_ JSON files.
If set to true a blank-line will be added at the end of a document. This is helpful when working with `git`, `diff` and 
other command line tools.

**Example:**

```xquery
serialize(
    map{ "a": 1 },
    map {
        "method": "json",
        "exist:insert-final-newline": true()
    }
)
```

will result in 

```json
{"a": 1}
\n
```

Note the newline at the end.

Tests: 
- XQSuite tests
- Tested to work in `fn:serialize`, `file:sync` (only XML) and `file:serialize`
- Adds first tests for `file:serialize` and `file:serialize-binary`. 

This PR also fixes a regression where only one Suite in XQuery3Tests was run.

### CODE REVIEW

**Comment by joewiz** [comment]
2 of the 3 remaining CI failures (both Windows) are identical to the failures in #4253:

CI: Test / (JDK 8 / windows-latest) Test (pull_request) 
Error: CollectionLocksTest.stress_noDeadlock_writeWrite_subtree:565->stress_noDeadlock:701 Stress test likely showed a thread deadlock
Links: [#4253](https://github.com/eXist-db/exist/runs/5462900543?check_suite_focus=true#step:5:8979) [this PR](https://github.com/eXist-db/exist/runs/5514286249?check_suite_focus=true#step:5:8989)

CI: (JDK 11 / windows-latest) Test
Error: "Multiple failures"
Links: [#4253](https://github.com/eXist-db/exist/runs/5462900624?check_suite_focus=true#step:5:17851) [this PR](https://github.com/eXist-db/exist/runs/5514286403?check_suite_focus=true)

As I noted in https://github.com/eXist-db/exist/pull/4253#issuecomment-1065570748, the fact that these very diferent PRs trip on the same errors in Windows suggests the issue isn't with these PRs per se, but that there's some existing issue with the develop branch in their interaction with the Windows environment.

The 3rd failure, unique to this PR, is in the macOS Test and is the one we've seen before intermittently with Maven's "dependency-check" and "OSS Index rate limit exceeded". So I think that can be disregarded.

**Review by joewiz — APPROVED** [review]
This works for me as advertised. Furthermore, it makes the new serialization method available via the `declare option exist:serialize` option:

```xquery
xquery version "3.1";

declare option exist:serialize "insert-final-newline=yes";

<x/>
```

--> returns `<x/>\n`

**Comment by line-o** [comment]
At the community call it was discussed if the line ending needs to be configurable or adhere to operating system settings.
I re-checked the implementation in `src/main/java/org/exist/util/serializer/IndentingXMLWriter.java` and conclude:
The line endings are hard-coded to `\n` throughout the class. It would therefore be bad advice to make it configurable at this point.

**Comment by adamretter** [comment]
@line-o The implementation is part of the Java Servlet specification (for us that is Eclipse Jetty) - https://docs.oracle.com/javaee/6/api/javax/servlet/http/HttpServletResponse.html#sendRedirect(java.lang.String)

**Comment by line-o** [comment]
Needs to be rebase once https://github.com/eXist-db/exist/pull/4290 was merged.

**Comment by joewiz** [comment]
@dependabot-bot rebase

**Comment by joewiz** [comment]
@dependabot rebase

**Comment by line-o** [comment]
@joewiz this needs to rebased manually. There are certain changes to be made before it can get merged.

**Comment by joewiz** [comment]
@line-o Just realized that! Sorry, I thought it would be ready once #4290 was ready. I will leave it to you to take remaining actions.

**Comment by line-o** [comment]
Manual rebase and commit history rewrite done.

**Review by dizzzz — APPROVED** [review]

File: exist-core/src/main/java/org/exist/storage/serializers/EXistOutputKeys.java
Comment (adamretter) [inline exist-core/src/main/java/org/exist/storage/serializers/EXistOutputKeys.java:116]: `public static final` - otherwise you will get static code complaints.

File: extensions/modules/file/src/test/xquery/modules/file/serialize.xqm
Comment (adamretter) [inline extensions/modules/file/src/test/xquery/modules/file/serialize.xqm:27]: Why are we using `resource:` here?

File: exist-core/src/main/java/org/exist/storage/serializers/EXistOutputKeys.java
Comment (line-o) [reply exist-core/src/main/java/org/exist/storage/serializers/EXistOutputKeys.java:116]: as discussed in the community call this should be done in a separate PR

File: extensions/modules/file/src/test/xquery/modules/file/serialize.xqm
Comment (line-o) [reply extensions/modules/file/src/test/xquery/modules/file/serialize.xqm:27]: As discussed in the community call this is necessary for the import to work.

File: exist-core/src/main/java/org/exist/storage/serializers/EXistOutputKeys.java
Comment (adamretter) [reply exist-core/src/main/java/org/exist/storage/serializers/EXistOutputKeys.java:116]: > as discussed in the community call this should be done in a separate PR

That is not my recollection of the discussion at all. My memory is that as this is new code it should be corrected.
You yourself even pointed out that you understood that others were opposed to fixing such issues in other large refactoring PRs.

File: extensions/modules/file/src/test/xquery/modules/file/serialize.xqm
Comment (adamretter) [reply extensions/modules/file/src/test/xquery/modules/file/serialize.xqm:27]: As discussed on the Community Call.. I will take a look into this to find out where the issue is.

Comment (line-o) [reply extensions/modules/file/src/test/xquery/modules/file/serialize.xqm:27]: Which will not block this PR getting merged right?
This method of importing modules is used in multiple places throughout the codebase.
I would prefer you open an issue and not discuss this here.

Comment (dizzzz) [reply extensions/modules/file/src/test/xquery/modules/file/serialize.xqm:27]: @adamretter ?

File: exist-core/src/main/java/org/exist/storage/serializers/EXistOutputKeys.java
Comment (line-o) [reply exist-core/src/main/java/org/exist/storage/serializers/EXistOutputKeys.java:116]: This is a one line addition to existing code where it follows the structure of the existing declarations. @dizzzz specifically said in the call that if you change one of the lines all of them should be changed, which in turn would obscure the PRs changes.
I am wondering if the order of keywords is blocking this PR.

**Review by dizzzz — APPROVED** [review]

**Comment by line-o** [comment]
rebasing this

**Comment by line-o** [comment]
OK, so now this should be re-reviewed and then hopefully merged.

**Comment by dizzzz** [comment]
Lets tak about the resource: item.

**Pull request merged**
